### PR TITLE
Add e-mail subscription changing from external service (Mailchimp)

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/Controller/Api/SubscriptionController.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * @Route("/email-subscriptions")
+ */
+class SubscriptionController extends Controller
+{
+    /**
+     * @Route("/change", name="app_change_email_subscriptions_webhook")
+     * @Method("POST")
+     */
+    public function changeEmailSubscriptionsAction(Request $request, SubscriptionHandler $subscriptionHandler): Response
```
Comment (HeahDude) [inline src/Controller/Api/SubscriptionController.php]: Please either declare this handler as public, or extends the `AbstractController` and declare it in `getSubscribedServices()`.

File: src/Controller/Api/SubscriptionController.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        if (!$request->query->has('secret') || !$secret = $request->query->get('secret')) {
+            return new Response('There is no secret.', Response::HTTP_UNAUTHORIZED);
```
Comment (HeahDude) [inline src/Controller/Api/SubscriptionController.php]:
```php
throw new UnauthorizedHttpException('There is no secret');
```

File: src/Controller/Api/SubscriptionController.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        if ($secret !== $this->container->getParameter('webhook.email_subscriptions_secret')) {
+            return new Response('Wrong secret.', Response::HTTP_UNAUTHORIZED);
```
Comment (HeahDude) [inline src/Controller/Api/SubscriptionController.php]:
```php
throw new AccessDeniedHttpException('Wrong secret.');
```

File: src/Controller/Api/SubscriptionController.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        $type = $request->request->get('type');
+        $data = $request->request->get('data');
+
+        if (0 === $request->request->count() || !$type || !isset($data['email']) || !isset($data['list_id'])) {
+            return new Response('The request should contain correct data.', Response::HTTP_BAD_REQUEST);
```
Comment (HeahDude) [inline src/Controller/Api/SubscriptionController.php]:
```php
throw new BadRequestHttpException('The request should contain correct data.');
```

File: src/Subscription/SubscriptionHandler.php [hunk 5]
```diff
@@ -3,23 +3,52 @@
...
+    private $subscriptionTypeRepository;
+
+    public function __construct(AdherentRepository $adherentRepository, SubscriptionTypeRepository $subscriptionTypeRepository)
     {
-        $this->repository = $repository;
+        $this->adherentRepository = $adherentRepository;
+        $this->subscriptionTypeRepository = $subscriptionTypeRepository;
+    }
+
+    public function changeSubscription(string $type, string $email, string $listId): void
```
Comment (HeahDude) [inline src/Subscription/SubscriptionHandler.php]: If type is not a vlid constant please throw an `\InvalidArgumentException`. And add a test case for it?

File: src/Subscription/SubscriptionHandler.php [hunk 6]
```diff
@@ -3,23 +3,52 @@
...
+    {
+        /** @var Adherent $adherent */
```
Comment (HeahDude) [inline src/Subscription/SubscriptionHandler.php]: Does the repo method use a type hint? You should be able to remove this.

File: src/Subscription/SubscriptionHandler.php [hunk 7]
```diff
@@ -3,23 +3,52 @@
...
+        $adherent = $this->adherentRepository->findOneByEmail($email);
+        if (!$adherent) {
+            throw new \RuntimeException(sprintf('There is no adherent with email address "%s".', $email));
+        }
+
+        /** @var SubscriptionType $subscriptionType */
+        $subscriptionType = $this->subscriptionTypeRepository->findOneBy(['externalId' => $listId]);
```
Comment (HeahDude) [inline src/Subscription/SubscriptionHandler.php]: Please add a custom method with proper type hint.

File: src/Subscription/SubscriptionHandler.php [hunk 8]
```diff
@@ -3,23 +3,52 @@
...
+        if (!$subscriptionType) {
+            throw new \RuntimeException(sprintf('There is no subscription type with external service id "%s".', $listId));
+        }
+
+        $hasSubscription = in_array($subscriptionType->getCode(), $adherent->getEmailsSubscriptions());
```
Comment (HeahDude) [inline src/Subscription/SubscriptionHandler.php]: Wouldn't it be better to use an adherent method for it? We might need this logic again elsewhere I guess. What about:
```php
$hasSubscription = $adherent->hasSubscriptionType($type);
```

File: src/Controller/Api/SubscriptionController.php (hunk 1 above)
Comment (illusionOfParadise) [reply src/Controller/Api/SubscriptionController.php]: Fixed.

File: src/Controller/Api/SubscriptionController.php (hunk 2 above)
Comment (illusionOfParadise) [reply src/Controller/Api/SubscriptionController.php]: Fixed.

File: src/Controller/Api/SubscriptionController.php (hunk 3 above)
Comment (illusionOfParadise) [reply src/Controller/Api/SubscriptionController.php]: Fixed.

File: src/Controller/Api/SubscriptionController.php (hunk 4 above)
Comment (illusionOfParadise) [reply src/Controller/Api/SubscriptionController.php]: Fixed.

File: src/Subscription/SubscriptionHandler.php (hunk 5 above)
Comment (illusionOfParadise) [reply src/Subscription/SubscriptionHandler.php]: Done.

File: src/Subscription/SubscriptionHandler.php (hunk 7 above)
Comment (illusionOfParadise) [reply src/Subscription/SubscriptionHandler.php]: Done.

File: src/Subscription/SubscriptionHandler.php (hunk 8 above)
Comment (illusionOfParadise) [reply src/Subscription/SubscriptionHandler.php]: Done.

File: src/Subscription/SubscriptionHandler.php (hunk 6 above)
Comment (illusionOfParadise) [reply src/Subscription/SubscriptionHandler.php]: Done.

**Pull request merged**